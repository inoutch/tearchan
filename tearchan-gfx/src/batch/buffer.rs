use crate::batch::BatchObjectId;
use crate::buffer::BufferInterface;
use std::collections::HashMap;
use tearchan_util::btree::DuplicatableBTreeMap;

const DEFAULT_BUFFER_SIZE: usize = 1024usize;
pub type BufferFactory<TBuffer> = fn(
    &<TBuffer as BufferInterface>::Device,
    &<TBuffer as BufferInterface>::Queue,
    &mut Option<&mut <TBuffer as BufferInterface>::Encoder>,
    Option<(&TBuffer, usize)>, // (previous buffer, previous buffer used size)
    usize,
) -> TBuffer;

#[derive(Debug, Clone)]
pub struct BatchPointer {
    pub first: usize,
    pub size: usize,
}

impl BatchPointer {
    pub fn new(first: usize, size: usize) -> Self {
        BatchPointer { first, size }
    }

    pub fn last(&self) -> usize {
        self.first + self.size
    }
}

pub struct BatchBuffer<TBuffer>
where
    TBuffer: BufferInterface,
{
    buffer: TBuffer,
    buffer_factory: BufferFactory<TBuffer>,
    pointers: HashMap<BatchObjectId, BatchPointer>,
    last: usize,
    flushed_last: usize,
    pending_pointers: DuplicatableBTreeMap<usize, BatchPointer>,
    fragmentation_size: usize,
}

impl<TBuffer> BatchBuffer<TBuffer>
where
    TBuffer: BufferInterface,
{
    pub fn new(
        device: &TBuffer::Device,
        queue: &TBuffer::Queue,
        encoder: &mut Option<&mut TBuffer::Encoder>,
        buffer_factory: BufferFactory<TBuffer>,
    ) -> Self {
        let factory = &buffer_factory;
        BatchBuffer {
            buffer: factory(device, queue, encoder, None, DEFAULT_BUFFER_SIZE),
            buffer_factory,
            pointers: HashMap::new(),
            last: 0,
            flushed_last: 0,
            pending_pointers: DuplicatableBTreeMap::default(),
            fragmentation_size: 0,
        }
    }

    pub fn allocate(
        &mut self,
        device: &TBuffer::Device,
        queue: &TBuffer::Queue,
        encoder: &mut Option<&mut TBuffer::Encoder>,
        id: BatchObjectId,
        size: usize,
    ) -> &mut BatchPointer {
        debug_assert!(!self.pointers.contains_key(&id));

        // Search from pending_pointers
        if let Some(mut ptr) = match self.pending_pointers.range_mut(size..).next() {
            Some((_, pointers)) => pointers.pop_back(),
            None => None,
        } {
            // Reuse the memory if there is more free space than the desired size
            self.fragmentation_size -= ptr.size; // Note that reduce will increase the fragment size
            if ptr.size != size {
                // Reducing unnecessary memory size
                self.reduce_pointer(queue, &mut ptr, size);
            }

            self.pointers.insert(id, ptr);
        } else {
            // Allocate new memory space
            let ptr = self.allocate_new_pointer(device, queue, encoder, size);
            self.pointers.insert(id, ptr);
        }
        self.pointers.get_mut(&id).unwrap()
    }

    pub fn reallocate(
        &mut self,
        device: &TBuffer::Device,
        queue: &TBuffer::Queue,
        encoder: &mut Option<&mut TBuffer::Encoder>,
        id: BatchObjectId,
        size: usize,
    ) {
        let mut pointer = self.pointers.remove(&id).unwrap();
        match pointer.size {
            d if d > size => {
                self.reduce_pointer(queue, &mut pointer, size);
                self.pointers.insert(id, pointer);
            }
            d if d < size => {
                self.buffer.clear(queue, pointer.first, pointer.size);
                self.fragmentation_size += pointer.size;

                self.allocate(device, queue, encoder, id, size);
                self.pending_pointers.push_back(pointer.size, pointer);
            }
            _ => {
                self.pointers.insert(id, pointer);
            }
        }
    }

    pub fn free(&mut self, queue: &TBuffer::Queue, id: BatchObjectId) {
        let pointer = self.pointers.remove(&id).unwrap();
        self.fragmentation_size += pointer.size;
        self.buffer.clear(queue, pointer.first, pointer.size);
        self.pending_pointers.push_back(pointer.size, pointer);
    }

    pub fn write(
        &self,
        queue: &TBuffer::Queue,
        pointer: &BatchPointer,
        data: &[TBuffer::DataType],
    ) {
        assert!(data.len() <= pointer.size);
        self.buffer
            .write(queue, bytemuck::cast_slice(data), pointer.first);
    }

    pub fn buffer(&self) -> &TBuffer {
        &self.buffer
    }

    pub fn len(&self) -> usize {
        self.buffer.len()
    }

    pub fn is_empty(&self) -> bool {
        self.buffer.is_empty()
    }

    pub fn last(&self) -> usize {
        self.last
    }

    pub fn fragmentation_size(&self) -> usize {
        self.fragmentation_size
    }

    // NOTICE: Destroy structures
    pub fn defragmentation(&mut self) {
        let mut first: usize = 0;
        for pointer in &mut self.pointers {
            pointer.1.first = first;
            first += pointer.1.size;
        }

        self.last = first;
        self.pending_pointers.clear();
        self.fragmentation_size = 0;
    }

    pub fn get_pointer(&self, id: &BatchObjectId) -> Option<&BatchPointer> {
        self.pointers.get(id)
    }

    pub fn sort(&mut self, ids: &[BatchObjectId]) {
        debug_assert_eq!(ids.len(), self.pointers.len());
        let mut pointers = HashMap::new();
        let mut size = 0;
        for id in ids {
            let pointer = &self.pointers[id];
            pointers.insert(*id, BatchPointer::new(size, pointer.size));
            size += pointer.size;
        }

        self.pointers = pointers;
        self.last = size;
    }

    fn reallocate_buffer(
        &mut self,
        device: &TBuffer::Device,
        queue: &TBuffer::Queue,
        encoder: &mut Option<&mut TBuffer::Encoder>,
        size: usize,
    ) {
        let new_size = size * 2;
        let factory = &self.buffer_factory;
        self.buffer = factory(
            device,
            queue,
            encoder,
            Some((&self.buffer, self.flushed_last)),
            new_size,
        );
    }

    fn allocate_new_pointer(
        &mut self,
        device: &TBuffer::Device,
        queue: &TBuffer::Queue,
        encoder: &mut Option<&mut TBuffer::Encoder>,
        size: usize,
    ) -> BatchPointer {
        let first = self.last;
        if first + size > self.buffer.len() {
            self.reallocate_buffer(device, queue, encoder, first + size);
        }

        self.last += size;
        BatchPointer::new(first, size)
    }

    fn reduce_pointer(&mut self, queue: &TBuffer::Queue, pointer: &mut BatchPointer, size: usize) {
        if pointer.last() != self.last {
            let r_first = pointer.first + size;
            let r_size = pointer.size - size;
            let r_ptr = BatchPointer::new(r_first, r_size);
            self.pending_pointers.push_back(r_size, r_ptr);

            self.buffer.clear(queue, r_first, r_size);
            self.fragmentation_size += r_size;
        } else {
            self.last = pointer.first + size;
        }

        pointer.size = size;
    }

    pub fn flush(&mut self) {
        // Update the size copied to the buffer.
        // This will determine how much of the existing buffer should be restored when the buffer is recreated.
        self.flushed_last = self.last;
    }
}

#[cfg(test)]
mod test {
    use crate::batch::buffer::BatchBuffer;
    use crate::buffer::BufferInterface;
    use std::cell::RefCell;

    struct MockBuffer(RefCell<Vec<u32>>);

    impl BufferInterface for MockBuffer {
        type DataType = u32;
        type Device = ();
        type Queue = ();
        type Encoder = ();

        fn write(&self, _queue: &Self::Queue, data: &[Self::DataType], offset: usize) {
            self.0.borrow_mut()[offset..(offset + data.len())].clone_from_slice(data);
        }

        fn len(&self) -> usize {
            self.0.borrow().len()
        }

        fn is_empty(&self) -> bool {
            self.0.borrow().is_empty()
        }

        fn clear(&self, _queue: &Self::Queue, offset: usize, len: usize) {
            self.0.borrow_mut()[offset..(offset + len)].fill(0);
        }
    }

    #[test]
    fn test_allocate() {
        let mut buffer: BatchBuffer<MockBuffer> =
            BatchBuffer::new(&(), &(), &mut None, |_, _, _, prev, len| {
                MockBuffer(RefCell::new(if let Some((prev_buffer, prev_len)) = prev {
                    let mut v = prev_buffer.0.borrow().clone();
                    for _ in prev_len..len {
                        v.push(0);
                    }
                    v
                } else {
                    vec![0; len]
                }))
            });
        let p0 = buffer.allocate(&(), &(), &mut None, 1, 10).clone();
        buffer.write(&(), &p0, &[1; 10]);
        assert_eq!(p0.first, 0);
        assert_eq!(p0.size, 10);

        let p1 = buffer.allocate(&(), &(), &mut None, 2, 15).clone();
        buffer.write(&(), &p1, &[2; 15]);
        assert_eq!(p1.first, 10);
        assert_eq!(p1.size, 15);

        let p2 = buffer.allocate(&(), &(), &mut None, 3, 5).clone();
        buffer.write(&(), &p2, &[3; 5]);
        assert_eq!(p2.first, 25);
        assert_eq!(p2.size, 5);

        let p3 = buffer.allocate(&(), &(), &mut None, 4, 12).clone();
        buffer.write(&(), &p3, &[4; 12]);
        assert_eq!(p3.first, 30);
        assert_eq!(p3.size, 12);

        buffer.flush();

        assert_eq!(
            &buffer.buffer.0.borrow()[0..buffer.last()],
            &vec![
                1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3,
                3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4
            ]
        );
    }

    #[test]
    fn test_free() {
        let mut buffer: BatchBuffer<MockBuffer> =
            BatchBuffer::new(&(), &(), &mut None, |_, _, _, prev, len| {
                MockBuffer(RefCell::new(if let Some((prev_buffer, prev_len)) = prev {
                    let mut v = prev_buffer.0.borrow().clone();
                    for _ in prev_len..len {
                        v.push(0);
                    }
                    v
                } else {
                    vec![0; len]
                }))
            });
        let p0 = buffer.allocate(&(), &(), &mut None, 1, 10).clone();
        buffer.write(&(), &p0, &[1; 10]);

        let p1 = buffer.allocate(&(), &(), &mut None, 2, 15).clone();
        buffer.write(&(), &p1, &[2; 15]);

        let p2 = buffer.allocate(&(), &(), &mut None, 3, 5).clone();
        buffer.write(&(), &p2, &[3; 5]);

        let p3 = buffer.allocate(&(), &(), &mut None, 4, 12).clone();
        buffer.write(&(), &p3, &[4; 12]);
        buffer.flush();

        buffer.free(&(), 1);
        buffer.free(&(), 2);
        buffer.free(&(), 3);
        buffer.free(&(), 4);
        assert_eq!(&buffer.buffer.0.borrow()[0..buffer.last()], &vec![0; 42]);

        let p4 = buffer.allocate(&(), &(), &mut None, 5, 16).clone();
        buffer.write(&(), &p4, &[5; 16]);
        assert_eq!(p4.first, 42);
        assert_eq!(p4.size, 16);

        let p5 = buffer.allocate(&(), &(), &mut None, 6, 12).clone();
        buffer.write(&(), &p5, &[6; 12]);
        assert_eq!(p5.first, 30);
        assert_eq!(p5.size, 12);

        let p6 = buffer.allocate(&(), &(), &mut None, 7, 5).clone();
        buffer.write(&(), &p6, &[7; 5]);
        assert_eq!(p6.first, 25);
        assert_eq!(p6.size, 5);

        assert_eq!(
            &buffer.buffer.0.borrow()[0..buffer.last()],
            &vec![
                0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 7, 7, 7,
                7, 7, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
                5, 5
            ]
        );
    }
}
