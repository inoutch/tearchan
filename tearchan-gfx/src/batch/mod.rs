use crate::batch::object::{BatchObject, BatchObjectCommand};
use crate::batch::types::{BatchAttributeIndex, BatchTypeArray, BatchTypeTransform};
use std::collections::hash_map::RandomState;
use std::collections::{HashMap, HashSet};
use std::sync::mpsc::{channel, Receiver, Sender};
use tearchan_util::btree::DuplicatableBTreeMap;
use tearchan_util::id_manager::{IdGenerator, IdManager};

const DEFAULT_ORDER: i32 = 0;

pub type BatchObjectId = u64;

pub mod batch2d;
pub mod batch3d;
pub mod batch_billboard;
pub mod batch_line;
pub mod buffer;
pub mod object;
pub mod types;
pub mod v2;

pub enum BatchProviderCommand<'a> {
    Add {
        id: BatchObjectId,
        data: &'a Vec<BatchTypeArray>,
        order: &'a Option<i32>,
    },
    Remove {
        id: BatchObjectId,
    },
    Replace {
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        data: &'a BatchTypeArray,
    },
}

pub trait BatchProvider {
    type Device;
    type Queue;
    type Encoder;

    fn run(
        &mut self,
        device: &Self::Device,
        queue: &Self::Queue,
        encoder: &mut Option<&mut Self::Encoder>,
        command: BatchProviderCommand,
    );

    fn sort(&mut self, ids: Vec<u64>) -> HashSet<u32, RandomState>;

    fn flush(&mut self, queue: &Self::Queue, manager: &mut BatchObjectManager);
}

pub struct BatchCommandBuffer {
    id_generator: IdGenerator<BatchObjectId>,
    command_sender: Sender<BatchObjectCommand>,
}

impl BatchCommandBuffer {
    pub fn add(&mut self, data: Vec<BatchTypeArray>, order: Option<i32>) -> BatchObjectId {
        let id = self.id_generator.gen();
        self.command_sender
            .send(BatchObjectCommand::Add { id, data, order })
            .unwrap();
        id
    }

    pub fn remove(&mut self, id: BatchObjectId) {
        self.command_sender
            .send(BatchObjectCommand::Remove { id })
            .unwrap();
    }

    pub fn replace(
        &mut self,
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        data: BatchTypeArray,
    ) {
        self.command_sender
            .send(BatchObjectCommand::Replace {
                id,
                attribute,
                data,
            })
            .unwrap();
    }

    pub fn transform(
        &mut self,
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        transform: BatchTypeTransform,
    ) {
        self.command_sender
            .send(BatchObjectCommand::Transform {
                id,
                attribute,
                transform,
            })
            .unwrap();
    }
}

pub struct Batch<T>
where
    T: BatchProvider,
{
    id_manager: IdManager<BatchObjectId>,
    command_receiver: Receiver<BatchObjectCommand>,
    command_sender: Sender<BatchObjectCommand>,
    batch_object_manager: BatchObjectManager,
    provider: T,
}

impl<T> Batch<T>
where
    T: BatchProvider,
{
    pub fn new(provider: T) -> Batch<T> {
        let (tx, rx) = channel();
        Batch {
            id_manager: IdManager::new(0, |id| *id + 1),
            command_receiver: rx,
            command_sender: tx,
            batch_object_manager: BatchObjectManager::default(),
            provider,
        }
    }
}

impl<T> Batch<T>
where
    T: BatchProvider,
{
    pub fn create_command_buffer(&mut self) -> BatchCommandBuffer {
        BatchCommandBuffer {
            id_generator: self.id_manager.create_generator(),
            command_sender: Sender::clone(&self.command_sender),
        }
    }

    pub fn flush(
        &mut self,
        device: &T::Device,
        queue: &T::Queue,
        encoder: &mut Option<&mut T::Encoder>,
    ) {
        let mut need_sort_all = false;
        while let Ok(command) = self.command_receiver.try_recv() {
            if let Some(provider_command) = match &command {
                BatchObjectCommand::Add { id, data, order } => {
                    if order.is_some() {
                        need_sort_all = true;
                    }
                    Some(BatchProviderCommand::Add {
                        id: *id,
                        data,
                        order,
                    })
                }
                BatchObjectCommand::Remove { id } => Some(BatchProviderCommand::Remove { id: *id }),
                BatchObjectCommand::Replace {
                    id,
                    attribute,
                    data,
                } => Some(BatchProviderCommand::Replace {
                    id: *id,
                    attribute: *attribute,
                    data,
                }),
                _ => None,
            } {
                self.provider.run(device, queue, encoder, provider_command);
            }
            self.batch_object_manager.run(command);
        }

        if need_sort_all {
            let attributes = self
                .provider
                .sort(self.batch_object_manager.create_sorted_ids());
            for attribute in attributes {
                self.batch_object_manager
                    .run(BatchObjectCommand::Refresh { attribute });
            }
        }

        self.provider.flush(queue, &mut self.batch_object_manager);
    }

    pub fn provider(&self) -> &T {
        &self.provider
    }
}

#[derive(Default)]
pub struct BatchObjectManager {
    objects: HashMap<BatchObjectId, BatchObject>,
    sorted_object_ids: DuplicatableBTreeMap<i32, BatchObjectId>,
    changed_object_attributes: HashMap<BatchObjectId, HashSet<BatchAttributeIndex>>,
}

impl BatchObjectManager {
    pub fn flush<F>(&mut self, mut callback: F)
    where
        F: FnMut(&mut BatchObject, u32),
    {
        for (id, attributes) in &mut self.changed_object_attributes {
            let object = match self.objects.get_mut(id) {
                Some(object) => object,
                None => continue,
            };
            for attribute in attributes.iter() {
                callback(object, *attribute);
            }
        }
        self.changed_object_attributes.clear();
    }

    fn run(&mut self, command: BatchObjectCommand) {
        match command {
            BatchObjectCommand::Add { id, data, order } => {
                self.command_add(id, data, order);
            }
            BatchObjectCommand::Remove { id } => {
                self.command_remove(id);
            }
            BatchObjectCommand::Transform {
                id,
                attribute,
                transform,
            } => {
                self.command_transform(id, attribute, transform);
            }
            BatchObjectCommand::Replace {
                id,
                data,
                attribute,
            } => {
                self.command_replace(id, data, attribute);
            }
            BatchObjectCommand::Refresh { attribute } => {
                self.command_refresh(attribute);
            }
        }
    }

    fn command_add(&mut self, id: BatchObjectId, data: Vec<BatchTypeArray>, order: Option<i32>) {
        let order = order.unwrap_or(DEFAULT_ORDER);
        let transforms = vec![BatchTypeTransform::None; data.len()];
        let mut set = HashSet::with_capacity(data.len());
        for i in 0u32..data.len() as u32 {
            set.insert(i);
        }

        self.objects
            .insert(id, BatchObject::new(id, data, transforms, order));
        self.sorted_object_ids.push_back(order, id);
        self.changed_object_attributes.insert(id, set);
    }

    fn command_remove(&mut self, id: BatchObjectId) {
        self.changed_object_attributes.remove(&id);
        if let Some(object) = self.objects.remove(&id) {
            self.sorted_object_ids.remove(&object.order(), &id);
        }
    }

    fn command_transform(
        &mut self,
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        transform: BatchTypeTransform,
    ) {
        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        object.set_transform(attribute, transform);

        if let Some(set) = self.changed_object_attributes.get_mut(&id) {
            set.insert(attribute);
        } else {
            let mut set = HashSet::new();
            set.insert(attribute);
            self.changed_object_attributes.insert(id, set);
        }
    }

    fn command_replace(
        &mut self,
        id: BatchObjectId,
        data: BatchTypeArray,
        attribute: BatchAttributeIndex,
    ) {
        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        object.set_data(attribute, data);

        self.refresh_attribute(id, attribute);
    }

    fn command_refresh(&mut self, attribute: BatchAttributeIndex) {
        let ids = self.objects.keys().into_iter().copied().collect::<Vec<_>>();
        for id in ids {
            self.refresh_attribute(id, attribute);
        }
    }

    fn refresh_attribute(&mut self, id: BatchObjectId, attribute: BatchAttributeIndex) {
        match self.changed_object_attributes.get_mut(&id) {
            None => {
                let mut set = HashSet::new();
                set.insert(attribute);
                self.changed_object_attributes.insert(id, set);
            }
            Some(set) => {
                set.insert(attribute);
            }
        }
    }

    fn create_sorted_ids(&self) -> Vec<BatchObjectId> {
        self.sorted_object_ids
            .iter()
            .map(|(_, ids)| ids)
            .flatten()
            .copied()
            .collect()
    }
}
