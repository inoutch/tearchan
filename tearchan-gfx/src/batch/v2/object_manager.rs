use crate::batch::types::{BatchAttributeIndex, BatchTypeArray, BatchTypeTransform};
use crate::batch::v2::buffer::{
    BatchBufferAllocator, BatchBufferAllocatorEvent, BatchBufferPointer,
};
use crate::batch::v2::object::BatchObject;
use crate::batch::DEFAULT_ORDER;
use std::collections::{HashMap, HashSet, VecDeque};
use tearchan_util::id_manager::IdManager;

#[derive(Hash, Eq, PartialEq, Copy, Clone, Debug)]
pub struct BatchObjectId(u64);

impl BatchObjectId {
    fn next(&self) -> Self {
        BatchObjectId(self.0 + 1)
    }
}

#[derive(Debug)]
pub enum BatchObjectEvent {
    Add {
        id: BatchObjectId,
    },
    Remove {
        id: BatchObjectId,
    },
    Write {
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
    },
    Clear {
        pointer: BatchBufferPointer,
    },
    Resize {
        len: usize,
    },
}

pub struct BatchObjectManager {
    id_manager: IdManager<BatchObjectId>,
    objects: HashMap<BatchObjectId, BatchObject>,
    object_ids_grouped_by_pointer: HashMap<BatchBufferPointer, BatchObjectId>,
    objects_will_be_rewritten: HashSet<(BatchObjectId, BatchAttributeIndex)>,
    allocator: BatchBufferAllocator,
    events: VecDeque<BatchObjectEvent>,
    len: usize,
}

impl BatchObjectManager {
    pub fn new(len: usize) -> Self {
        BatchObjectManager {
            id_manager: IdManager::new(BatchObjectId(0), |id| id.next()),
            objects: HashMap::new(),
            objects_will_be_rewritten: HashSet::new(),
            object_ids_grouped_by_pointer: HashMap::new(),
            allocator: BatchBufferAllocator::default(),
            events: VecDeque::new(),
            len,
        }
    }

    pub fn pop_event(&mut self) -> Option<BatchObjectEvent> {
        while let Some(event) = self.allocator.pop_event() {
            match event {
                BatchBufferAllocatorEvent::Write(pointer) => {
                    let object_id = self.object_ids_grouped_by_pointer.get(&pointer).unwrap();
                    let object = self.objects.get(object_id).unwrap();
                    for (i, _) in object.data().iter().enumerate() {
                        self.events.push_back(BatchObjectEvent::Write {
                            id: *object_id,
                            attribute: i as BatchAttributeIndex,
                        });
                    }
                }
                BatchBufferAllocatorEvent::Clear(pointer) => {
                    self.events.push_back(BatchObjectEvent::Clear { pointer });
                }
                BatchBufferAllocatorEvent::Reallocate { from, to } => {
                    let object_id = *self.object_ids_grouped_by_pointer.get(&from).unwrap();
                    let object = self.objects.get_mut(&object_id).unwrap();
                    self.object_ids_grouped_by_pointer.remove(&object.pointer());

                    object.set_pointer(to);
                    self.object_ids_grouped_by_pointer.insert(to, object_id);

                    for i in 0u32..object.data().len() as u32 {
                        self.objects_will_be_rewritten.insert((object_id, i));
                    }
                }
            }
        }
        let event = self.events.pop_front()?;
        match &event {
            BatchObjectEvent::Add { .. } => {}
            BatchObjectEvent::Remove { .. } => {}
            BatchObjectEvent::Write { id, attribute } => {
                self.objects_will_be_rewritten.remove(&(*id, *attribute));
            }
            BatchObjectEvent::Clear { .. } => {}
            BatchObjectEvent::Resize { .. } => {}
        }
        Some(event)
    }

    pub fn clear_events(&mut self) {
        self.objects_will_be_rewritten.clear();
    }

    #[inline]
    pub fn get(&self, id: BatchObjectId) -> Option<&BatchObject> {
        self.objects.get(&id)
    }

    pub fn add(
        &mut self,
        data: Vec<BatchTypeArray>,
        len: usize,
        order: Option<i32>,
    ) -> BatchObjectId {
        for datum in data.iter() {
            assert_eq!(datum.len(), len);
        }

        let id = self.id_manager.gen();
        let order = order.unwrap_or(DEFAULT_ORDER);
        let transforms = vec![BatchTypeTransform::None; data.len()];
        let pointer = self.allocator.allocate(len);
        let object = BatchObject::new(pointer, data, transforms, order);

        if self.allocator.len() > self.len {
            self.len = self.allocator.len() * 2;
            self.events
                .push_back(BatchObjectEvent::Resize { len: self.len });
        }

        for i in 0u32..object.data().len() as u32 {
            self.objects_will_be_rewritten.insert((id, i));
        }
        self.events.push_back(BatchObjectEvent::Add { id });
        self.objects.insert(id, object);
        self.object_ids_grouped_by_pointer.insert(pointer, id);
        id
    }

    pub fn remove(&mut self, id: BatchObjectId) -> Option<BatchObject> {
        self.events.push_back(BatchObjectEvent::Remove { id });
        let object = self.objects.remove(&id)?;
        self.allocator.free(object.pointer());
        self.object_ids_grouped_by_pointer.remove(&object.pointer());
        for attribute in 0..object.data().len() {
            self.objects_will_be_rewritten
                .remove(&(id, attribute as u32));
        }
        Some(object)
    }

    pub fn transform(
        &mut self,
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        transform: BatchTypeTransform,
    ) {
        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        object.set_transform(attribute, transform);

        let key = (id, attribute);
        if self.objects_will_be_rewritten.get(&key).is_none() {
            self.events
                .push_back(BatchObjectEvent::Write { id, attribute });
            self.objects_will_be_rewritten.insert(key);
        }
    }

    pub fn rewrite(
        &mut self,
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        data: BatchTypeArray,
    ) {
        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        let prev_data = match object.data().get(attribute as usize) {
            Some(data) => data,
            None => return,
        };
        assert_eq!(data.len(), prev_data.len());

        object.set_data(attribute, data);
        let key = (id, attribute);
        if self.objects_will_be_rewritten.get(&key).is_none() {
            self.events
                .push_back(BatchObjectEvent::Write { id, attribute });
            self.objects_will_be_rewritten.insert(key);
        }
    }

    pub fn replace(&mut self, id: BatchObjectId, len: usize, data: Vec<BatchTypeArray>) {
        for datum in data.iter() {
            assert_eq!(datum.len(), len);
        }

        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        if !object.data().is_empty() && object.data().get(0).unwrap().len() == len {
            for (i, data) in data.into_iter().enumerate() {
                self.rewrite(id, i as u32, data);
            }
            return;
        }

        self.object_ids_grouped_by_pointer.remove(&object.pointer());

        let new_pointer = self.allocator.reallocate(object.pointer(), len);
        object.set_pointer(new_pointer);
        self.object_ids_grouped_by_pointer.insert(new_pointer, id);

        for i in 0u32..object.data().len() as u32 {
            self.objects_will_be_rewritten.insert((id, i));
        }
    }

    pub fn allocator_len(&self) -> usize {
        self.allocator.len()
    }

    pub fn allocator_is_empty(&self) -> bool {
        self.allocator.is_empty()
    }
}

#[cfg(test)]
mod test {
    use crate::batch::types::{BatchTypeArray, BatchTypeTransform};
    use crate::batch::v2::object_manager::{BatchObjectEvent, BatchObjectManager};
    use nalgebra_glm::{vec2, vec3, Mat2, Mat3};

    fn convert_events(manager: &mut BatchObjectManager) -> Vec<BatchObjectEvent> {
        let mut events = Vec::new();
        while let Some(event) = manager.pop_event() {
            events.push(event);
        }
        events
    }

    #[test]
    fn test_valid_len() {
        let mut manager = BatchObjectManager::new(100);
        manager.add(
            vec![
                BatchTypeArray::V1F32 { data: vec![0.0f32] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(0.0f32, 0.0f32)],
                },
            ],
            1,
            None,
        );
    }

    #[test]
    #[should_panic]
    fn test_invalid_len() {
        let mut manager = BatchObjectManager::new(100);
        manager.add(
            vec![
                BatchTypeArray::V1F32 {
                    data: vec![0.0f32, 0.0f32],
                },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(0.0f32, 0.0f32)],
                },
            ],
            2,
            None,
        );
    }

    #[test]
    fn test_cleanup() {
        let mut manager = BatchObjectManager::new(100);
        let data = vec![
            BatchTypeArray::V1F32 { data: vec![0.0f32] },
            BatchTypeArray::V2F32 {
                data: vec![vec2(0.0f32, 0.0f32)],
            },
        ];
        let id0 = manager.add(data, 1, None);

        let data = vec![
            BatchTypeArray::V1F32 { data: vec![0.0f32] },
            BatchTypeArray::V2F32 {
                data: vec![vec2(0.0f32, 0.0f32)],
            },
        ];
        let id1 = manager.add(data, 1, None);
        manager.remove(id0);
        manager.remove(id1);

        assert_eq!(manager.objects.len(), 0);
        assert_eq!(manager.object_ids_grouped_by_pointer.len(), 0);
        assert_eq!(manager.objects_will_be_rewritten.len(), 0);
        assert_eq!(manager.allocator.len(), 0);
    }

    #[test]
    fn test_events() {
        let mut manager = BatchObjectManager::new(100);
        let id0 = manager.add(
            vec![
                BatchTypeArray::V1F32 { data: vec![1.0f32] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(1.0f32, 1.0f32)],
                },
            ],
            1,
            None,
        );
        let id1 = manager.add(
            vec![
                BatchTypeArray::V1F32 { data: vec![2.0f32] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(2.0f32, 2.0f32)],
                },
            ],
            1,
            None,
        );

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        manager.transform(
            id0,
            1,
            BatchTypeTransform::Mat3F32 {
                m: Mat3::new(
                    1.0f32, 0.0f32, 0.0f32, 0.0f32, 1.0f32, 0.0f32, 0.0f32, 0.0f32, 1.0f32,
                ),
            },
        );
        manager.transform(
            id1,
            0,
            BatchTypeTransform::Mat2F32 {
                m: Mat2::new(1.0f32, 0.0f32, 0.0f32, 1.0f32),
            },
        );

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        manager.remove(id0);

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        manager.remove(id1);

        insta::assert_debug_snapshot!(convert_events(&mut manager));
    }

    #[test]
    fn test_resize_event() {
        let mut manager = BatchObjectManager::new(2);
        let _id0 = manager.add(
            vec![
                BatchTypeArray::V1U32 { data: vec![1, 1] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(1.0f32, 1.0f32), vec2(1.0f32, 1.0f32)],
                },
                BatchTypeArray::V3F32 {
                    data: vec![vec3(1.0f32, 1.0f32, 1.0f32), vec3(1.0f32, 1.0f32, 1.0f32)],
                },
            ],
            2,
            None,
        );

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        let _id1 = manager.add(
            vec![
                BatchTypeArray::V1U32 {
                    data: vec![2, 2, 2],
                },
                BatchTypeArray::V2F32 {
                    data: vec![
                        vec2(2.0f32, 2.0f32),
                        vec2(2.0f32, 2.0f32),
                        vec2(2.0f32, 2.0f32),
                    ],
                },
                BatchTypeArray::V3F32 {
                    data: vec![
                        vec3(2.0f32, 2.0f32, 2.0f32),
                        vec3(2.0f32, 2.0f32, 2.0f32),
                        vec3(2.0f32, 2.0f32, 2.0f32),
                    ],
                },
            ],
            3,
            None,
        );
        insta::assert_debug_snapshot!(convert_events(&mut manager));
    }
}
