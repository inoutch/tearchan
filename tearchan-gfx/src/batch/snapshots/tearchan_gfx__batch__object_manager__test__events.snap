---
source: tearchan-gfx/src/batch/v2/object_manager.rs
expression: convert_events(&mut manager)

---
[
    Add {
        id: BatchObjectId(
            0,
        ),
    },
    Add {
        id: BatchObjectId(
            1,
        ),
    },
    WriteToIndexBuffer {
        id: BatchObjectId(
            0,
        ),
    },
    WriteToIndexBuffer {
        id: BatchObjectId(
            1,
        ),
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            0,
        ),
        attribute: 0,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            0,
        ),
        attribute: 1,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            1,
        ),
        attribute: 0,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            1,
        ),
        attribute: 1,
    },
]
